### ISSUE TITLE
[pro] Stable cursor when ordering by nullable columns in PostgreSQL

### ISSUE THREAD

**Issue by reppie** [rmosolgo/graphql-ruby#2284]
We recently tried to use GraphQL::Pro's stable cursor, but ran into problems when sorting by nullable fields in our postgres setup.

The query that is generated includes a `WHERE #{column} = NULL` or a `WHERE #{column} > NULL`, which never returns any results. 

We managed to make `ordered_relation.rb#build_condition` return correct queries by:
- Adding `OR #{column_id} IS NULL` to all cases of the if/else conditions
- Adding a `if value.nil?` check that determines whether to use `IS` or `=`

This isn't a pretty solution by any means, as it will add the `OR #{column_id} IS NULL` condition in all cases, i.e. even on primary keys.

What would be a good/better way to address this problem?

**Comment by rmosolgo** [comment]
Hey, sorry for the late response here. I was away for a new baby and the catch-up is ... ongoing 😅 . 

It seems like your workaround of checking for `nil` and adding the `IS NULL` syntax is a pretty good solution. Is there any reason I shouldn't patch graphql-pro the same way?

**Comment by reppie** [comment]
Hey, no worries about the delay :)

Yeah it was a solid workaround, just that we weren't a fan of the fact that `OR #{column_id} IS NULL` was added to all the conditions in the query. We couldn't really think of a neater fix though, so patching it the same way would work too yeah.

**Comment by rmosolgo** [comment]
I took a look at this, but I'm not sure I really _can_ support `NULL`s with stable cursors. The thing is, when I'm in "deep" into `NULL` territory, I can't figure out how to work my way out. What I mean is, imagine a relation like this, sorted by `name asc, id asc`: 

name | id 
---|---
Alpha Romeo | 3 
Bugatti | 1 
Chevrolet  | 4
NULL | 2
NULL | 5 
NULL | 6 

Let's say you've worked part way in and you're asking for 

```ruby 
cars(first: 2, after: "#{cursor_from("Chevrolet", 4)}")  { ...  }
```

(cursors left in plaintext for example purposes) 

That would give you two cursors: 

- `cursor_from(nil, 2)` 
- `cursor_from(nil, 5)` 

Then, lets say you want to go _back_ from that: 

```ruby 
cars(last: 2, before: "#{cursor_from(nil, 5)}") { ... }
``` 

That's going to make a SQL query roughly like this: 

```SQL
SELECT ... WHERE ((name < NULL) OR (name IS NULL AND id < 5))
```

Of course, `name < NULL` isn't going to work, because that's always false. Another option would be to use `name IS NOT NULL`, but that doesn't work because it includes _all three_ of the preceding entries. 

So, I can't figure out: once the cursor is _based on_ `nil`, how can we _reenter_ the range of the relation with proper values? Is there a solution here I've overlooked? 

One possibility would be to add the last non-`nil` value to the cursor, which we could use when backwards-paginating. That would be really complicated -- the cursor would contain a value _not_ belonging to the current record, and we'd have to do some detection on when to use or ignore it. 

### Workarounds? 

I was hoping to find a work-around, I tried two things but neither one worked. 

First, I tried ordering by `COALESCE`, for example `.order('COALESCE(name, "zzzzz")')`. But this didn't work because, in order to get _backwards_ pagination checks (eg `hasPreviousPage`), we call `ActiveRecord::Relation#reverse_order`. But, this ordering is too complicated for ActiveRecord to reverse, so it :boom:s with `IrreversibleOrderError`.

Second, I tried adding `COALESE` to the `SELECT`, then ordering by it. But the problem there is, the cursor fields are _re-added_ to the `SELECT`, and it ends up putting `SELECT ... cars.name as cursor_0` which doesn't work 😖 

I'm not really sure where to go from here. One or more of: 

- Add partial support for sorting by `NULL`, document the limitations and try to rescue the ActiveRecord errors that might pop up. 
- Don't add support, document that it must be ordered by non-null columns

😩 

Could you tell me more about your use case? Is there a chance that a "zero-value" (like `0` or `""`) would be appropriate in your data model?

**Comment by willglynn** [comment]
> Of course, `name < NULL` isn't going to work, because that's always false. Another option would be to use `name IS NOT NULL`, but that doesn't work because it includes _all three_ of the preceding entries.

I don't understand. Isn't that normal?

The usual backwards-paginating SQL is of the form:

```sql
WHERE (name < ?) OR (name = ? AND id < ?) ORDER BY name DESC, id DESC LIMIT n
```

This `WHERE` matches all the records before this one, and it uses `ORDER BY … LIMIT` to return just the next page. Match every record with this name and a preceding ID, also match every record with a preceding name, sort them and return just the next page.

Going backwards from `null` seems equivalent:

```sql
WHERE (name IS NOT NULL) OR (name IS NULL AND id < ?) ORDER BY name DESC, id DESC LIMIT n
```

Match every record with this name (`NULL`) and a preceding ID, also match every record with a preceding (non-`NULL`) name, sort them and return just the next page.

**Comment by rmosolgo** [comment]
Thanks for that suggestion. I agree it could work _if you know_ that `name IS NOT NULL` is equivalent to `name > NULL` (that is, the database is sorting nulls last). The problem is, databases handle nulls differently: 

- Postgres: nulls greater than any value, reordered with `NULLS FIRST | LAST`
- SQLite: nulls less than any value
- MySQL: nulls less than any value, unless `ORDER BY -...`

So, detecting whether nulls should be "greater" or "less" than a provided value (or whether actual values should be "greater" or "less" than `NULL`) is tricky.

(I realized this by making it work for SQLite ... then it was broken on Postgres.)

I _think_ I could implement it on a backend-by-backend basis; that is, make a Postgres-only stable connection wrapper (which handles nulls in postgres's way) and then another one for the other DBs (maybe they could share one).

**Comment by rmosolgo** [comment]
Well, it's finally out there 🎉 . I've just released graphql-pro 1.12.x which includes a new stable connection implementation with better null handling on per-database basis: 

https://graphql-ruby.org/pagination/stable_relation_connections.html

It requires graphql-ruby 1.10.x, though, since it builds on the improved pagination system in that version. 

If you're able to run on those versions and try it out, please let me know how it goes for you! If not, please let me know whatever trouble you run into in updating and I'll be happy to see what I can do. Thanks a ton for your patience here!
